File: samples/vision/yolov5/runtime/python/yolov5.py
# Copyright (c) 2025 D-Robotics Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# flake8: noqa: E501
# flake8: noqa: E402

import os
import cv2
import sys
import hbm_runtime
import numpy as np
from dataclasses import dataclass, field
from typing import Optional, Dict, Tuple, Literal

# Add project root to sys.path so we can import utility modules.
# Source files:
#   utils/py_utils/preprocess_utils.py
#   utils/py_utils/postprocess_utils.py
# (Check these files for preprocessing/postprocessing implementations.)
sys.path.append(os.path.abspath("../../../../../"))
import utils.py_utils.preprocess as pre_utils
import utils.py_utils.postprocess as post_utils


def sigmoid(x: np.ndarray) -> np.ndarray:
    """
    @brief Compute the sigmoid activation function.
    @param x Input NumPy array.
    @return NumPy array after applying sigmoid function element-wise.
    """
    return 1.0 / (1.0 + cv2.exp(-x))


def decode_layer(feat: np.ndarray,
                 stride: int,
                 anchor: np.ndarray,
                 classes_num: int = 80) -> np.ndarray:
    """
    @brief Decode a single feature layer from detection head.
    @param feat Raw model output tensor of shape (1, na, h, w, c).
    @param stride Stride of the feature layer.
    @param anchor Anchor sizes for this layer (na, 2).
    @param classes_num Number of output classes.
    @return Decoded prediction array of shape (N, 5 + num_classes).
    """
    _, _, h, w, _ = feat.shape  #  h/w: feature map size

    # Create coordinate grid of shape (1, 1, h, w, 2)
    grid_y, grid_x = np.mgrid[0:h, 0:w]
    grid = np.stack((grid_x, grid_y), axis=-1)[None, None]

    # batch sigmoid
    feat_sig = sigmoid(feat[..., :5 + classes_num])

    # Decode center offsets (dx, dy) and size (dw, dh)
    dxdy = feat_sig[..., :2]
    dwdh = feat_sig[..., 2:4]
    obj  = feat_sig[..., 4:5]
    cls  = feat_sig[..., 5:]

    # Compute center coordinates in original image scale
    xy = (dxdy * 2. - 0.5 + grid) * stride

    # Compute width/height from anchor sizes
    wh = (dwdh * 2.) ** 2 * anchor[:, None, None, :]

    # Construct final output tensor (xywh + obj + class scores)
    out = np.empty((*xy.shape[:-1], 5 + classes_num), dtype=np.float32)
    out[..., 0:2] = xy
    out[..., 2:4] = wh
    out[..., 4:5] = obj
    out[..., 5:]  = cls

    return out.reshape(-1, 5 + classes_num)


def decode_outputs(output_names: list[str],
                   fp32_outputs: dict[str, np.ndarray],
                   strides: list[int],
                   anchors: list[np.ndarray],
                   classes_num: int = 80) -> np.ndarray:
    """
    @brief Decode all feature maps from model output.
    @param output_names List of output tensor names.
    @param fp32_outputs Dict of decoded tensors from model.
    @param strides Stride values for each output head.
    @param anchors Anchor arrays for each head.
    @param classes_num Number of output classes.
    @return Concatenated prediction tensor of shape (N, 5 + classes).
    """
    decoded = []
    for i, key in enumerate(output_names):
        out = fp32_outputs[key]
        h, w = out.shape[1:3]
        # Reshape and transpose to (1, na, h, w, c)
        feat = out.reshape(1, h, w, 3, 5 + classes_num).transpose(0, 3, 1, 2, 4)
        decoded.append(decode_layer(feat, strides[i], anchors[i], classes_num))
    return np.concatenate(decoded, axis=0)


@dataclass
class YOLOv5Config:
    """
    @brief Configuration for initializing YoloV5X model.

    Contains model path and all runtime parameters required for preprocessing,
    inference, and postprocessing.

    Fields:
        model_path (str):
            Path to the compiled YOLOv5 .hbm model.

        classes_num (int):
            Number of detection categories. Default: 80.

        resize_type (int):
            Image resize mode used during preprocessing.
            0 = stretch resize
            1 = keep aspect ratio with padding

        score_thres (float):
            Minimum confidence threshold for filtering candidates.

        nms_thres (float):
            IoU threshold used for Non-Maximum Suppression.

        strides (np.ndarray):
            Feature map strides, default [8,16,32].

        anchors (np.ndarray):
            YOLO anchor definitions with shape (3, 3, 2).
    """
    model_path: str
    classes_num: int = 80
    resize_type: int = 1
    score_thres: float = 0.25
    nms_thres: float = 0.45
    # Feature map downsampling strides
    strides: np.ndarray = field(
        default_factory=lambda: np.array([8, 16, 32], dtype=np.int32)
    )
    # Anchors for each scale
    anchors: np.ndarray = field(
        default_factory=lambda: np.array([
            [10, 13], [16, 30], [33, 23],
            [30, 61], [62, 45], [59, 119],
            [116, 90], [156, 198], [373, 326]
        ], dtype=np.float32).reshape(3, 3, 2)
    )


class YoloV5X:
    """
    @brief YOLOv5X object detection wrapper using HB_HBMRuntime.

    Provides a unified inference pipeline including input preprocessing,
    model execution, and postprocessing (decode, confidence filtering,
    and Non-Maximum Suppression).
    """

    def __init__(self, config: YOLOv5Config):
        """
        @brief Initialize the YOLOv5X model with a provided configuration.

        @param config (YOLOv5Config)
            Configuration object containing model path, preprocessing,
            and postprocessing parameters. All field descriptions and
            constraints are defined within the YOLOv5Config dataclass.
        """

        # Load model and extract metadata
        self.model = hbm_runtime.HB_HBMRuntime(config.model_path)

        self.model_name = self.model.model_names[0]
        self.input_names = self.model.input_names[self.model_name]
        self.output_names = self.model.output_names[self.model_name]
        self.input_shapes = self.model.input_shapes[self.model_name]
        self.output_quants = self.model.output_quants[self.model_name]

        # Model input resolution (H, W) inferred from model input tensor
        self.input_h = self.input_shapes[self.input_names[0]][1]
        self.input_w = self.input_shapes[self.input_names[0]][2]

        # Detection and preprocessing configuration
        self.cfg = config

    def set_scheduling_params(self,
                              priority: Optional[int] = None,
                              bpu_cores: Optional[list] = None) -> None:
        """
        @brief Configure inference scheduling parameters.

        @param priority (int, optional)
            Inference priority in range [0, 255].
        @param bpu_cores (list[int], optional)
            BPU core indices used for inference.
        @return None
        """
        kwargs = {}
        if priority is not None:
            kwargs["priority"] = {self.model_name: priority}
        if bpu_cores is not None:
            kwargs["bpu_cores"] = {self.model_name: bpu_cores}

        if kwargs:
            self.model.set_scheduling_params(**kwargs)

    def pre_process(self, img: np.ndarray,
                    resize_type: Optional[int] = None,
                    image_format: Optional[str] = "BGR"
                    ) -> Dict[str, Dict[str, np.ndarray]]:
        """
        @brief Preprocess input image into model-required tensor format.

        The input image is resized according to resize_type and converted
        from BGR to NV12 (Y + UV planes).

        @param img (np.ndarray)
            Input image array.
        @param resize_type (int, optional)
            Resize strategy. If None, uses the value from configuration.
            If provided, the configuration value will be updated.
        @param image_format (str, optional)
            Input image format. Currently only "BGR" is supported.
        @return dict
            Nested input tensor dictionary:
            {model_name: {input_name: tensor}}
        """
        if resize_type == None:
            resize_type = self.cfg.resize_type
        else:
            self.cfg.resize_type = resize_type

        # Resize and convert to NV12
        if(image_format == "BGR"):
            resize_img = pre_utils.resized_image(img, self.input_w, self.input_h, resize_type)
            y, uv = pre_utils.bgr_to_nv12_planes(resize_img)
        else :
            raise ValueError(f"Unsupported image_format: {image_format}")

        return {
            self.model_name: {
                self.input_names[0]: y,
                self.input_names[1]: uv
            }
        }

    def forward(self, input_tensor: Dict[str, Dict[str, np.ndarray]]) -> Dict[str, np.ndarray]:
        """
        @brief Execute model inference.

        @param input_tensor (dict)
            Preprocessed input tensor dictionary produced by pre_process().
        @return dict
            Raw output tensors returned by the runtime.
        """
        outputs = self.model.run(input_tensor)
        return outputs

    def post_process(self,
                     outputs: Dict[str, Dict[str, np.ndarray]],
                     ori_img_w: int,
                     ori_img_h: int,
                     score_thres: Optional[float] = None,
                     nms_thres: Optional[float] = None,
                     ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        @brief Convert raw model outputs into final detection results.

        This includes dequantization, decoding, confidence filtering,
        Non-Maximum Suppression (NMS), and coordinate scaling.

        @param outputs (dict)
            Raw output tensors from inference.
        @param ori_img_w (int)
            Width of the original input image.
        @param ori_img_h (int)
            Height of the original input image.
        @param score_thres (float, optional)
            Confidence threshold. If None, uses configuration value.
        @param nms_thres (float, optional)
            NMS IoU threshold. If None, uses configuration value.
        @return Tuple:
            - xyxy (np.ndarray): Bounding boxes (N, 4) in original image coordinates.
            - cls (np.ndarray): Class indices (N).
            - score (np.ndarray): Confidence scores (N).
        """
        score_thres = score_thres or self.cfg.score_thres
        nms_thres = nms_thres or self.cfg.nms_thres

        # Step 1: Convert quantized outputs to float32
        fp32_outputs = post_utils.dequantize_outputs(outputs[self.model_name], self.output_quants)

        # Step 2: Decode YOLO outputs into unified predictions
        pred = decode_outputs(self.output_names, fp32_outputs,
                              self.cfg.strides, self.cfg.anchors, self.cfg.classes_num)

        # Step 3: Filter predictions by confidence threshold
        xyxy_boxes, score, cls = post_utils.filter_predictions(pred, score_thres)

        # Step 4: Non-Maximum Suppression (NMS)
        keep = post_utils.NMS(xyxy_boxes, score, cls, nms_thres)

        # Step 5: Rescale boxes to original image dimensions
        xyxy = post_utils.scale_coords_back(xyxy_boxes[keep], ori_img_w, ori_img_h,
                                            self.input_w, self.input_h, self.cfg.resize_type)

        return xyxy, score[keep], cls[keep]

    def predict(self,
                img: np.ndarray,
                image_format: str = "BGR",
                resize_type: Optional[int] = None,
                score_thres: Optional[float] = None,
                nms_thres: Optional[float] = None,
                ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        @brief Run a complete detection pipeline on a single image.

        This method internally performs preprocessing, inference,
        and postprocessing.

        @param img (np.ndarray)
            Input image array.
        @param image_format (str)
            Input image format. Currently supports "BGR".
        @param resize_type (int, optional)
            Resize strategy override.
        @param score_thres (float, optional)
            Confidence threshold override.
        @param nms_thres (float, optional)
            NMS IoU threshold override.
        @return Tuple:
            - xyxy (np.ndarray): Bounding boxes (N, 4).
            - cls (np.ndarray): Class indices (N).
            - score (np.ndarray): Confidence scores (N).
        """
        # Original image size
        ori_img_h, ori_img_w = img.shape[:2]

        # 1) Preprocess
        input_tensor = self.pre_process(img, resize_type, image_format)

        # 2) Inference
        outputs = self.forward(input_tensor)

        # 3) Postprocess
        xyxy, score, cls = self.post_process(outputs, ori_img_w, ori_img_h,
                                             score_thres, nms_thres)

        return xyxy, score, cls

    def __call__(self,
                 img: np.ndarray,
                 image_format: str = "BGR",
                 resize_type: Optional[int] = None,
                 score_thres: Optional[float] = None,
                 nms_thres: Optional[float] = None,
                 ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        @brief Callable interface for the detection pipeline.

        Equivalent to calling predict().

        @param img (np.ndarray)
            Input image.
        @param image_format (str)
            Input image format.
        @return Same as predict().
        """
        return self.predict(img, image_format, resize_type, score_thres, nms_thres)

File: samples/vision/yolov5/runtime/cpp/src/yolov5.cc
/*
 * Copyright (c) 2025, XiangshunZhao D-Robotics.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file yolov5.cc
 * @brief Implement the YOLOv5x inference pipeline using HB-DNN / UCP runtime APIs.
 *
 *        This file contains the concrete implementation of the YOLOv5x wrapper
 *        and its end-to-end inference flow on Horizon Robotics platforms:
 *        - Initialize and load a packed *.hbm model, query tensor properties,
 *          and allocate stride-aware tensor buffers.
 *        - Preprocess input images (letterbox resize and BGR->NV12 conversion)
 *          and write data into input tensor memory with proper cache handling.
 *        - Execute synchronous inference tasks on BPU via hbDNNInferV2() and UCP
 *          scheduling APIs, then invalidate output caches for CPU-side access.
 *        - Postprocess raw outputs by dequantization, decoding, thresholding,
 *          NMS, and coordinate scaling back to the original image space.
 *
 *        The implementation is designed as a reference pipeline that separates
 *        preprocessing, runtime execution, and postprocessing into reusable steps
 *        while keeping resource ownership and cleanup explicit.
 */


#include "yolov5.hpp"
#include <chrono>

/**
 * @brief Construct a YOLOv5x instance.
 *
 * The object is created in an uninitialized state. Call init() to load the
 * model, query tensor properties, and allocate tensor buffers.
 */
YOLOv5x::YOLOv5x()
{
    // Number of models contained in the packed DNN handle
    model_count = 0;

    // Packed DNN handle (may contain multiple models)
    packed_dnn_handle = nullptr;

    // Handle of the YOLOv5x model used for inference
    dnn_handle = nullptr;

    // Number of input tensors required by the model
    input_count = 0;

    // Number of output tensors produced by the model
    output_count = 0;

    // Model input height (pixels)
    input_h = 0;

    // Model input width (pixels)
    input_w = 0;

    inited = false;
}

/**
 * @brief Initialize model resources from a *.hbm model file.
 *
 * This function performs the actual initialization, including:
 * - Loading the packed model from disk
 * - Selecting the first model handle in the pack
 * - Querying input/output tensor properties
 * - Allocating input/output tensor buffers
 * - Caching model input resolution (H/W)
 *
 * Calling init() more than once is not allowed.
 *
 * @param model_path [in] Path to the quantized *.hbm model file.
 * @return int32_t 0 on success, non-zero on failure.
 */
int32_t YOLOv5x::init(const char* model_path)
{
    const char **model_name_list = nullptr;

    if (inited) {
        fprintf(stderr, "YOLOv5x::init() called twice\n");
        return -1;
    }

    // Initialize DNN from file(s)
    HBDNN_CHECK_SUCCESS(hbDNNInitializeFromFiles(&packed_dnn_handle, &model_path, 1),
                        "hbDNNInitializeFromFiles failed");

    // Query available model names in the pack
    HBDNN_CHECK_SUCCESS(hbDNNGetModelNameList(&model_name_list, &model_count, packed_dnn_handle),
                        "hbDNNGetModelNameList failed");

    // Use the first model in the pack
    HBDNN_CHECK_SUCCESS(hbDNNGetModelHandle(&dnn_handle, packed_dnn_handle, model_name_list[0]),
                        "hbDNNGetModelHandle failed");

    // Query input and output tensor counts
    HBDNN_CHECK_SUCCESS(hbDNNGetInputCount(&input_count, dnn_handle),
                        "hbDNNGetInputCount failed");
    HBDNN_CHECK_SUCCESS(hbDNNGetOutputCount(&output_count, dnn_handle),
                        "hbDNNGetOutputCount failed");

    // Prepare tensor descriptor arrays
    input_tensors.resize(input_count);
    output_tensors.resize(output_count);

    // Clear sysMem fields so destructor can safely free memory
    for (int i = 0; i < input_count; ++i) {
        std::memset(&input_tensors[i].sysMem, 0, sizeof(hbUCPSysMem));
    }

    for (int i = 0; i < output_count; ++i) {
        std::memset(&output_tensors[i].sysMem, 0, sizeof(hbUCPSysMem));
    }

    // Fetch input tensor properties
    for (int i = 0; i < input_count; i++) {
        HBDNN_CHECK_SUCCESS(hbDNNGetInputTensorProperties(&input_tensors[i].properties, dnn_handle, i),
                            "hbDNNGetInputTensorProperties failed");
    }

    // Fetch output tensor properties
    for (int i = 0; i < output_count; i++) {
        HBDNN_CHECK_SUCCESS(hbDNNGetOutputTensorProperties(&output_tensors[i].properties, dnn_handle, i),
                            "hbDNNGetOutputTensorProperties failed");
    }

    // Cache model input resolution
    input_h = input_tensors[0].properties.validShape.dimensionSize[1];
    input_w = input_tensors[0].properties.validShape.dimensionSize[2];

    // Allocate memory for all tensors
    prepare_input_tensor(input_tensors);
    prepare_output_tensor(output_tensors);

    inited = true;

    return 0;
}

/**
* @brief Destructor.
*
* Releases allocated tensor memory and DNN resources.
* Safe to call even if init() failed partially.
*/
YOLOv5x::~YOLOv5x()
{
    // Free input tensor memory
    for (int i = 0; i < input_count; i++) {
        hbUCPFree(&(input_tensors[i].sysMem));
    }

    // Free output tensor memory
    for (int i = 0; i < output_count; i++) {
        hbUCPFree(&(output_tensors[i].sysMem));
    }

    // Release packed model handle
    if (packed_dnn_handle != nullptr) {
        hbDNNRelease(packed_dnn_handle);
        packed_dnn_handle = nullptr;
    }
}

/**
 * @brief Preprocess an input image into model input tensor buffers.
 *
 * Current implementation:
 * - Letterbox resize to model input resolution (input_w/input_h)
 * - Convert BGR image to NV12 as required by the compiled model
 * - Write the converted data into input_tensors[0].sysMem
 *
 * @param input_tensors [in,out] Model input tensors to be filled.
 * @param input_w [in] Model input width in pixels.
 * @param input_h [in] Model input height in pixels.
 * @param img [in] Input image (OpenCV Mat).
 * @param image_format [in] Input image format string (only "BGR" is supported).
 *
 * @return int32_t 0 on success, non-zero on failure.
 */
int32_t pre_process(std::vector<hbDNNTensor>& input_tensors,
                    cv::Mat& img,
                    const int input_w, const int input_h,
                    const std::string& image_format)
{
    if (input_tensors[0].sysMem.virAddr == nullptr) {
        fprintf(stderr, "YOLOv5x not initialized.\n");
        return -1;
    }

    if (image_format == "BGR")
    {
        // Letterbox resize to model input resolution
        cv::Mat resized_mat;
        resized_mat.create(input_h, input_w, img.type());
        letterbox_resize(img, resized_mat);

        // Convert BGR to NV12 and write into input tensors
        bgr_to_nv12_tensor(resized_mat, input_tensors, input_h, input_w);
    }
    else
    {
        std::cerr << "[YOLOv5x::pre_process] Unsupported image_format: "
                  << image_format
                  << ". Supported formats: BGR."
                  << std::endl;
        return -1;
    }

    return 0;
}

/**
 * @brief Execute model inference on BPU.
 *
 * This function:
 * - Creates inference tasks via hbDNNInferV2()
 * - Submits it to the scheduler
 * - Waits for task to complete (blocking)
 * - Invalidates output tensor cache for CPU access
 * - Releases both task handle
 *
 * @param output_tensors [in,out] Output tensors to be filled by runtime.
 * @param input_tensors [in] Prepared model input tensors.
 * @param dnn_handle [in] DNN model handle used for inference.
 * @param sched_param [in] Optional scheduling parameters (nullptr uses default HB_UCP_BPU_CORE_ANY).
 *
 * @return int32_t 0 on success, non-zero on failure.
 */
int32_t infer(std::vector<hbDNNTensor>& output_tensors,
              std::vector<hbDNNTensor>& input_tensors,
              const hbDNNHandle_t dnn_handle,
              hbUCPSchedParam* sched_param)
{
    hbUCPTaskHandle_t task_handle{nullptr};

    // Create inference task
    HBDNN_CHECK_SUCCESS(hbDNNInferV2(&task_handle, output_tensors.data(), input_tensors.data(), dnn_handle),
                        "hbDNNInferV2 failed");


    // Prepare scheduling parameters
    hbUCPSchedParam local_param;
    hbUCPSchedParam* param_to_use = nullptr;

    if (sched_param == nullptr) {
        HB_UCP_INITIALIZE_SCHED_PARAM(&local_param);
        local_param.backend = HB_UCP_BPU_CORE_ANY;
        param_to_use = &local_param;
    } else {
        param_to_use = sched_param;
    }

    param_to_use->priority = 0;
    HBUCP_CHECK_SUCCESS(hbUCPSubmitTask(task_handle, param_to_use),
                        "hbUCPSubmitTask failed");

    // Wait until task completion (blocking)
    HBUCP_CHECK_SUCCESS(hbUCPWaitTaskDone(task_handle, 0),
                        "hbUCPWaitTaskDone failed");

    // Invalidate output tensor cache for CPU access
    for (int i = 0; i < output_tensors.size(); i++) {
        hbUCPMemFlush(&output_tensors[i].sysMem, HB_SYS_MEM_CACHE_INVALIDATE);
    }

    // Release task handle
    HBUCP_CHECK_SUCCESS(hbUCPReleaseTask(task_handle), "hbUCPReleaseTask failed");

    return 0;
}

/**
 * @brief Postprocess model outputs into final detection results.
 *
 * Steps:
 * - Dequantize each output tensor to float (current implementation uses S32 dequantization)
 * - Decode bounding boxes and class scores
 * - Apply confidence threshold filtering (config.score_thresh)
 * - Perform class-wise Non-Maximum Suppression (NMS) (config.nms_thresh)
 * - Map bounding boxes back to the original image space (letterbox inverse)
 *
 * @param results [out] Final detection results after postprocessing.
 * @param output_tensors [in] Raw output tensors from inference.
 * @param config [in] Postprocess configuration (anchors/strides/hw_list/thresholds/classes, etc.).
 * @param orig_img_w [in] Width of the original input image.
 * @param orig_img_h [in] Height of the original input image.
 * @param input_w [in] Model input width.
 * @param input_h [in] Model input height.
 */
void post_process(std::vector<Detection>& results,
                  std::vector<hbDNNTensor>& output_tensors,
                  const Yolov5Config& config,
                  int orig_img_w, int orig_img_h,
                  int input_w, int input_h)
{
    std::vector<std::vector<float>> float_outputs;

    // Dequantize each output tensor to float
    for (int i = 0; i < output_tensors.size(); ++i) {
        const auto& tensor = output_tensors[i];
        auto deq_result = dequantizeTensorS32(tensor);
        float_outputs.push_back(std::move(deq_result));
    }

    // Decode all detection layers
    auto decode_all = yolov5_decode_all_layers(
        float_outputs,
        config.hw_list,
        config.strides,
        config.anchors,
        config.score_thresh);

    // Apply class-wise NMS
    results = nms_bboxes(decode_all, config.nms_thresh);

    // Map bounding boxes back to original image coordinates
    scale_letterbox_bboxes_back(results, orig_img_w, orig_img_h, input_w, input_h);
}
